src/Lean/Server/InfoUtils.lean

```diff
@@ -189,4 +189,41 @@ where
       cs.any (hasNestedTactic pos tailPos)
     | _ => false
 
+/--
+Find info nodes that should be used for the term goal feature.
+
+The main complication concerns applications
+like `f a b` where `f` is an identifier.
+In this case, the term goal at `f`
+should be the goal for the full application `f a b`.
+
+Therefore we first gather the position of
+these head function symbols such as `f`,
+and later ignore identifiers at these positions.
+-/
+partial def InfoTree.termGoalAt? (t : InfoTree) (hoverPos : String.Pos) : Option (ContextInfo × Info) :=
+  let headFns : Std.HashSet String.Pos := t.foldInfo (init := {}) fun ctx i headFns => do
+    if let some pos := getHeadFnPos? i.stx then
+      headFns.insert pos
+    else
+      headFns
+
+  t.smallestInfo? fun i =>
+    if let (some pos, some tailPos) := (i.pos?, i.tailPos?) then
+      if pos ≤ hoverPos ∧ hoverPos < tailPos then
+        match i with
+        | Info.ofTermInfo ti => !ti.stx.isIdent || !headFns.contains pos
+        | _ => false
+      else false
+    else false
+
+  where
```

Comment:
We don't usually add empty lines around `where`, which makes it a bit easier to visually scan for declarations.